File: backend/app/services/project_service.py
"""Project service."""

from app.models import Project, ProjectCreate, ProjectUpdate, ProjectWithStats
from app.repositories import ProjectRepository, TaskRepository


class ProjectService:
    """Service for project operations."""

    def __init__(self):
        self.project_repo = ProjectRepository()
        self.task_repo = TaskRepository()

    def get_all(self) -> list[Project]:
        """Get all projects."""
        return self.project_repo.find_all()

    def get_by_id(self, project_id: int) -> Project | None:
        """Get project by ID."""
        return self.project_repo.find_by_id(project_id)

    def get_with_stats(self, project_id: int) -> ProjectWithStats | None:
        """Get project with task statistics."""
        project = self.project_repo.find_by_id(project_id)
        if not project:
            return None

        counts = self.task_repo.count_by_project(project_id)

        return ProjectWithStats(
            **project.model_dump(),
            task_count=sum(counts.values()),
            todo_count=counts.get("TODO", 0),
            in_progress_count=counts.get("IN_PROGRESS", 0),
            done_count=counts.get("DONE", 0),
        )

    def create(self, project_data: ProjectCreate) -> Project:
        """Create a new project."""
        return self.project_repo.create(project_data)

    def update(self, project_id: int, project_data: ProjectUpdate) -> Project | None:
        """Update a project."""
        return self.project_repo.update(project_id, project_data)

    def delete(self, project_id: int) -> bool:
        """Delete a project and its tasks."""
        # Delete associated tasks first
        self.task_repo.delete_by_project(project_id)
        return self.project_repo.delete(project_id)

    def exists(self, project_id: int) -> bool:
        """Check if project exists."""
        return self.project_repo.find_by_id(project_id) is not None

File: backend/src/services/projectService.js
/**
 * Project Service
 */
import { ProjectRepository } from '../repositories/projectRepository.js';
import { TaskRepository } from '../repositories/taskRepository.js';

class ProjectService {
  constructor() {
    this.projectRepo = new ProjectRepository();
    this.taskRepo = new TaskRepository();
  }

  /**
   * Get all projects
   * @returns {Promise<Array>} All projects
   */
  async getAll() {
    return await this.projectRepo.findAll();
  }

  /**
   * Get project by ID
   * @param {number} projectId - Project ID
   * @returns {Promise<object|null>} Project or null
   */
  async getById(projectId) {
    return await this.projectRepo.findById(projectId);
  }

  /**
   * Get project with task statistics
   * @param {number} projectId - Project ID
   * @returns {Promise<object|null>} Project with stats or null
   */
  async getWithStats(projectId) {
    return await this.projectRepo.findByIdWithStats(projectId);
  }

  /**
   * Create a new project
   * @param {object} projectData - Project data
   * @returns {Promise<object>} Created project
   */
  async create(projectData) {
    return await this.projectRepo.create(projectData);
  }

  /**
   * Update a project
   * @param {number} projectId - Project ID
   * @param {object} projectData - Project data to update
   * @returns {Promise<object|null>} Updated project or null
   */
  async update(projectId, projectData) {
    return await this.projectRepo.update(projectId, projectData);
  }

  /**
   * Delete a project and its tasks
   * @param {number} projectId - Project ID
   * @returns {Promise<boolean>} True if deleted
   */
  async delete(projectId) {
    // Delete associated tasks first
    await this.taskRepo.deleteByProject(projectId);
    return await this.projectRepo.delete(projectId);
  }

  /**
   * Check if project exists
   * @param {number} projectId - Project ID
   * @returns {Promise<boolean>} True if exists
   */
  async exists(projectId) {
    return await this.projectRepo.exists(projectId);
  }
}

export { ProjectService };
